File: src/main/java/al132/alchemistry/recipes/CombinerRecipe.java
package al132.alchemistry.recipes;

import al132.alchemistry.Ref;
import al132.alchemistry.utils.IItemHandlerUtils;
import al132.alchemistry.utils.StackUtils;
import al132.alib.tiles.CustomStackHandler;
import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.items.IItemHandler;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class CombinerRecipe {
    public final ItemStack output;
    public final List<ItemStack> inputs = new ArrayList<>();
    public final Set<Integer> nonEmptyIndices = new HashSet<>();

    //String gamestage;

    public CombinerRecipe(ItemStack output, List<Object> inObjs) {
        this.output = output;
        for (int i = 0; i < 9; i++) {
            Object temp;
            if (inObjs.size() > i) temp = inObjs.get(i);
            else temp = null;
            if (temp instanceof ItemStack) {
                inputs.add(((ItemStack) temp).copy());
            } else if (temp instanceof Item) {
                inputs.add(new ItemStack((Item) temp));
            } else if (temp instanceof Block) {
                inputs.add(new ItemStack((Block) temp));
            } else {
                inputs.add(ItemStack.EMPTY);
            }
            if (!inputs.get(i).isEmpty()) nonEmptyIndices.add(i);
        }
        assert inputs.size() == 9;
    }

    public List<ItemStack> getEmptyStrippedInputs() {
        return inputs.stream().filter(x -> !x.isEmpty()).collect(Collectors.toList());
    }


    public static CombinerRecipe matchInputs(IItemHandler handler) {
        return matchInputs(IItemHandlerUtils.toStackList(handler));//.toStackList())
    }

    private static CombinerRecipe matchInputs(List<ItemStack> inputStacks) {
        outer:
        for (CombinerRecipe recipe : ModRecipes.combinerRecipes) {
            int matchingStacks = 0;
            inner:
            for (int index = 0; index < recipe.inputs.size(); index++) {
                //inner: for ((index: Int, recipeStack: ItemStack) in recipe.inputs.withIndex()) {
                ItemStack recipeStack = recipe.inputs.get(index);
                ItemStack inputStack = inputStacks.get(index);
                if ((inputStack.getItem() == Ref.slotFiller || inputStack.isEmpty()) && recipeStack.isEmpty()) {
                    continue inner;
                } else if (!(StackUtils.areStacksEqualIgnoreQuantity(inputStack, recipeStack)
                        && inputStack.getCount() >= recipeStack.getCount())) {
                    // && (inputStack.get == recipeStack.itemDamage || recipeStack.itemDamage == OreDictionary.WILDCARD_VALUE))) {
                    continue outer;
                } else if (inputStack.isEmpty() || recipeStack.isEmpty()) {
                    continue outer;
                }
            }
            return recipe;//.copy()
        }
        return null;
    }

    public boolean matchesHandlerStacks(CustomStackHandler handler) {
        int matchingStacks = 0;

        for (int index = 0; index < this.inputs.size(); index++) {
            ItemStack recipeStack = this.inputs.get(index);
            ItemStack handlerStack = handler.getStackInSlot(index);
            if ((handlerStack.getItem() == Ref.slotFiller || handlerStack.isEmpty()) && recipeStack.isEmpty()) {
                matchingStacks++;
            } else if (handlerStack.isEmpty() || recipeStack.isEmpty()) continue;
            else if (StackUtils.areStacksEqualIgnoreQuantity(handlerStack, recipeStack)
                    && handlerStack.getCount() >= recipeStack.getCount()) {
                //&& (handlerStack.itemDamage == recipeStack.itemDamage || recipeStack.itemDamage == OreDictionary.WILDCARD_VALUE)) {
                matchingStacks++;
            }
        }
        return matchingStacks == 9;
    }

    public static CombinerRecipe matchOutput(ItemStack stack) {
        return ModRecipes.combinerRecipes.stream()
                .filter(it -> it.output.getItem() == stack.getItem())
                .filter(it -> ItemStack.areItemStacksEqual(it.output, stack))
                .findFirst()
                .orElse(null);
    }
}

File: src/main/java/al132/alchemistry/recipes/CombinerRecipe.kt
package al132.alchemistry.recipes

import al132.alchemistry.items.ModItems
import al132.alib.tiles.ALTileStackHandler
import al132.alib.utils.Utils.areItemsEqualIgnoreMeta
import al132.alib.utils.extensions.areItemStacksEqual
import al132.alib.utils.extensions.get
import al132.alib.utils.extensions.toStackList
import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.oredict.OreDictionary
import java.util.*

/**
 * Created by al132 on 1/22/2017.
 */

data class CombinerRecipe(val output: ItemStack, private val objsIn: List<Any?>, var gamestage: String = "") {

    val inputs = ArrayList<ItemStack>()

    init {
        val tempInputs = objsIn
        (0 until INPUT_COUNT).forEach { index ->
            val tempInput = tempInputs.getOrNull(index)
            when (tempInput) {
                is ItemStack -> inputs.add(tempInput)
                is Item      -> inputs.add(ItemStack(tempInput))
                is Block     -> inputs.add(ItemStack(tempInput))
                else         -> inputs.add(ItemStack.EMPTY)
            }
        }
    }

    fun matchesHandlerStacks(handler: ALTileStackHandler): Boolean {
        var matchingStacks = 0

        for ((index: Int, recipeStack: ItemStack) in this.inputs.withIndex()) {
            val handlerStack = handler[index]
            if ((handlerStack.item == ModItems.slotFiller || handlerStack.isEmpty) && recipeStack.isEmpty) matchingStacks++
            else if (handlerStack.isEmpty || recipeStack.isEmpty) continue
            else if (areItemsEqualIgnoreMeta(handlerStack, recipeStack)
                    && handlerStack.count >= recipeStack.count
                    && (handlerStack.itemDamage == recipeStack.itemDamage || recipeStack.itemDamage == OreDictionary.WILDCARD_VALUE)) {
                matchingStacks++
            }
        }
        return (matchingStacks == CombinerRecipe.INPUT_COUNT)
    }

    companion object {

        private const val INPUT_COUNT = 9

        fun matchInputs(handler: IItemHandler): CombinerRecipe? {
            assert(handler.slots == INPUT_COUNT)
            return matchInputs(handler.toStackList())
        }

        private fun matchInputs(inputStacks: List<ItemStack>): CombinerRecipe? {
            outer@ for (recipe in ModRecipes.combinerRecipes) {
                var matchingStacks = 0
                inner@ for ((index: Int, recipeStack: ItemStack) in recipe.inputs.withIndex()) {
                    val inputStack: ItemStack = inputStacks[index]
                    if ((inputStack.item == ModItems.slotFiller || inputStack.isEmpty) && recipeStack.isEmpty) {
                        continue@inner
                    } else if (!(areItemsEqualIgnoreMeta(inputStack, recipeStack)
                                    && inputStack.count >= recipeStack.count
                                    && (inputStack.itemDamage == recipeStack.itemDamage || recipeStack.itemDamage == OreDictionary.WILDCARD_VALUE))) {
                        continue@outer;
                    } else if (inputStack.isEmpty || recipeStack.isEmpty) {
                        continue@outer
                    }
                }
                return recipe//.copy()
            }
            return null
        }

        fun matchOutput(stack: ItemStack): CombinerRecipe? {
            return ModRecipes.combinerRecipes
                    .filter { it.output.item == stack.item }
                    .firstOrNull { it.output.areItemStacksEqual(stack) }
        }
    }
}